README.md

```diff
@@ -1358,6 +1359,142 @@ attributes_to_string(attrs)
+
+### Using `{% provide %}` tag
+
+First we use the `{% provde %}` tag to define the data we want to "provide" (make available).
+
+```django
+{% provide "my_data" key="hi" another=123 %}
+    {% component "child" %}  <--- Can access "my_data"
+    {% endcomponent %}
+{% endprovide %}
+
+{% component "child" %}  <--- Cannot access "my_data"
+{% endcomponent %}
+```
+
+Notice that the `provide` tag REQUIRES a name as a first argument. This is the _key_ by which we can then access the data passed to this tag.
+
+`provide` tag _key_, similarly to the _name_ argument in `component` or `slot` tags, has these requirements:
+- The _key_ must be a string literal
+- It must be a valid identifier (AKA a valid Python variable name)
+
+Once you've set the name, you define the data you want to "provide" by passing it as keyword arguments. This is similar to how you pass data to the `{% with %}` tag.
+
+> NOTE: Kwargs passed to `{% provide %}` are NOT added to the context.
+> In the example below, the `{{ key }}` won't render anything:
+> ```django
+> {% provide "my_data" key="hi" another=123 %}
+>     {{ key }} 
+> {% endprovide %}
+> ```
+
+### Using `inject()` method
+
+To "inject" (access) the data defined on the `provide` tag, you can use the `inject()` method inside of `get_context_data()`.
+
+For a component to be able to "inject" some data, the component (`{% component %}` tag) must be nested inside the `{% provide %}` tag.
+
+In the example from previous section, we've defined two kwargs: `key="hi" another=123`. That means that if we now inject `"my_data"`, we get an object with 2 attributes - `key` and `another`.
+
+```py
+class ChildComponent(component.Component):
+    def get_context_data(self):
+        my_data = self.inject("my_data")
+        print(my_data.key)     # hi
+        print(my_data.another) # 123
+        return {}
+```
+
+First argument to `inject` is the _key_ of the provided data. This
+must match the string that you used in the `provide` tag. If no provider
+with given key is found, `inject` raises a `KeyError`.
+
+To avoid the error, you can pass a second argument to `inject` to which will act as a default value, similar to `dict.get(key, default)`:
+
+```py
+class ChildComponent(component.Component):
+    def get_context_data(self):
+        my_data = self.inject("invalid_key", DEFAULT_DATA)
+        assert my_data == DEFAUKT_DATA
+        return {}
+```
+
+
+> NOTE: `inject()` works strictly only in `get_context_data`. If you try to call it from elsewhere, it will raise an error.
+
+#### (Im)mutability
+
+The instance returned from `inject()` is a subclass of `NamedTuple`, so the instance is immutable. This ensures that the data returned from `inject` will always
+have all the keys that were passed to the `provide` tag.
+
+However, while you CANNOT mutate the top-level instance:
+
+`self.inject("foo").var = 123`
+
+You CAN mutate the data nested under individual keys:
+
+`self.inject("foo").a_list.append(123)`
```

Comment:
That's a good point! Now that I'm thinking about it, mutating the nested data should be done sparingly, and those who will need it will figure it out. If explictly documented like before, it could encourage people to misuse.